_posts/2018-08-04-pfsense-as-hvm-on-pc-engines-apu2.md

```diff
@@ -0,0 +1,574 @@
+libxl: error: libxl_create.c:1255:domcreate_launch_dm: unable to add disk devices
+libxl: error: libxl_device.c:1086:device_backend_callback: unable to remove device with path /local/domain/0/backend/vbd/1/51712
+libxl: error: libxl.c:1647:devices_destroy_cb: libxl__devices_destroy failed for 1
+libxl: error: libxl.c:1575:libxl__destroy_domid: non-existant domain 1
+libxl: error: libxl.c:1534:domain_destroy_callback: unable to destroy guest with domid 1
+libxl: error: libxl.c:1463:domain_destroy_cb: destruction of domain 1 failed
+```
+
+# Crash after couple tries
+
+After couple tries of creating pfSense VM I faced below error:
+
+```
+root@apu2:~# xl create pfsense.cfg
+Parsing config from pfsense.cfg
+libxl: error: libxl_exec.c:118:libxl_report_child_exitstatus: /etc/xen/scripts/block add [490] exited with error status 1
+libxl: error: libxl_device.c:1237:device_hotplug_child_death_cb: script: Failed to find an unused loop device
+libxl: error: libxl_create.c:1255:domcreate_launch_dm: unable to add disk devices
+libxl: error: libxl_exec.c:118:libxl_report_child_exitstatus: /etc/xen/scripts/block remove [604] exited with error status 1
+libxl: error: libxl_device.c:1237:device_hotplug_child_death_cb: script: /etc/xen/scripts/block failed; error detected.
+libxl: error: libxl.c:1575:libxl__destroy_domid: non-existant domain 1
+libxl: error: libxl.c:1534:domain_destroy_callback: unable to destroy guest with domid 1
+libxl: error: libxl.c:1463:domain_destroy_cb: destruction of domain 1 failed
+```
+
+Solution: recompile kernel with `BLK_DEV_LOOP`
+
+# Read-only not supported
+
+```
+root@apu2:~# xl create pfsense.cfg
+Parsing config from pfsense.cfg
+libxl: error: libxl_dm.c:1433:libxl__build_device_model_args_new: qemu-xen doesn't support read-only IDE disk drivers
+libxl: error: libxl_dm.c:2182:device_model_spawn_outcome: (null): spawn failed (rc=-6)
+libxl: error: libxl_create.c:1504:domcreate_devmodel_started: device model did not start: -6
+libxl: error: libxl.c:1575:libxl__destroy_domid: non-existant domain 1
+libxl: error: libxl.c:1534:domain_destroy_callback: unable to destroy guest with domid 1
+libxl: error: libxl.c:1463:domain_destroy_cb: destruction of domain 1 failed
+```
+
+Solution: change pfsense.cfg by adding `rw` to img file.
+
+# References
+
+* [Install pfSense 2.1 RC0 amd64 on Xen 4.3 as PV HVM](https://forum.netgate.com/topic/58662/howto-install-pfsense-2-1-rc0-amd64-on-xen-4-3-as-pv-hvm)
+* [Virtual firewall](https://en.wikipedia.org/wiki/Virtual_firewall)
+* [xl.cfg manual for Xen 4.8](https://xenbits.xen.org/docs/4.8-testing/man/xl.cfg.5.html)
+* [Unanswered question about DMA attacks](https://security.stackexchange.com/questions/176503/dma-attacks-despite-iommu-isolation)
+* [Google blog post about fuzzying PCIe](https://cloudplatform.googleblog.com/2017/02/fuzzing-PCI-Express-security-in-plaintext.html)
+
+# Summary
+
+I hope this post was useful for you. Please feel free to share you opinion and
```

Comment:
Please feel free to share your opinion and...

Reply:
done